stage/stageqry.man
.\" @(#)$RCSfile: stageqry.man,v $ $Revision: 1.16 $ $Date: 2001/09/26 09:13:55 $ CERN IT-PDP/DM Jean-Philippe Baud
.\" Copyright (C) 1994-1999 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH STAGEQRY l "$Date: 2001/09/26 09:13:55 $"
.SH NAME
stageqry \- give the status of all staged tapes
.SH SYNOPSIS
.B stageqry
[
.BI -a
] [
.BI -h " host"
] [
.BI -L
] [
.BI -l
] [
.BI -P
] [
.BI -p " pool"
] [
.BI -q " file_sequence_number"
] [
.BI -Q " file_sequence_range"
] [
.BI -S
] [
.BI -T
] [
.BI -u
] [
.BI -V " visual_identifier(s)"
] [
.BI -x
] [
.BI --dump
] [
.BI --mintime
] [
.BI --reqid " reqid "
] [
.BI --retenp
]
.LP
.B stageqry
[
.BI -a
] [
.BI -f
] [
.BI -h " host"
] [
.BI -I " external_filename"
] [
.BI -L
] [
.BI -l
] [
.BI -P
] [
.BI -p " pool"
] [
.BI -S
] [
.BI -u
] [
.BI -x
] [
.BI --dump
] [
.BI --mintime
] [
.BI --reqid " reqid "
] [
.BI --retenp
]
.LP
.B stageqry
[
.BI -a
] [
.BI -f
] [
.BI -h " host"
] [
.BI -L
] [
.BI -l
] [
.BI -M " hsmfile | pattern"
] [
.BI -P
] [
.BI -p " pool"
] [
.BI -S
] [
.BI -u
] [
.BI -x
] [
.BI --dump
] [
.BI --mintime
] [
.BI --noregexp
] [
.BI --reqid " reqid "
] [
.BI --retenp
]
.LP
.B stageqry
[
.BI -A " pattern"
] [
.BI -a
] [
.BI -f
] [
.BI -h " host"
] [
.BI -L
] [
.BI -l
] [
.BI -P
] [
.BI -p " pool"
] [
.BI -S
] [
.BI -u
] [
.BI -x
] [
.BI --dump
] [
.BI --mintime
] [
.BI --noregexp
] [
.BI --reqid " reqid "
] [
.BI --retenp
]
.LP
.B stageqry
[
.BI -h " host"
] [
.BI -p " pool"
] [
.BI -s
] [
.BI --class
] [
.BI --counters
] [
.BI --mintime
] [
.BI --migrator
] [
.BI --queue
] [
.BI --retenp
]
.SH DESCRIPTION
.B stageqry
gives the status of all tapes which are staged or currently requested
in tabular form:
.HP
Vid	visual identifier.
.HP
Fseq	file sequence number.
.HP
Lbl	label type:
.BR al ,
.BR nl ,
.B sl
or
.BR blp .
.HP
Recfm	record format:
.BR F ,
.B FB
or
.BR U .
.TP
Lrecl	logical record length.
.HP
Blksiz	maximum block size.
.HP
State	current status:
.RS
STAGEALLOC	space has been reserved with stagealloc
.TP
STAGEIN	currently reading from tape to disk
.TP
STAGEOUT	space has been reserved for stageout file
.TP
STAGEWRT	currently copying from disk to tape
.TP
WAITING_SPC	waiting for space
.TP
WAITING_REQ	waiting for the same request by another job to complete
.TP
STAGED	successfully staged to EOF
.TP
STAGED_LSZ	successfully staged to specified size
.TP
STAGED_TPE	successfully staged but blocks were skipped
.TP
PUT_FAILED	stageput failed
.TP
CAN_BE_MIGR	candidate for the next automatic migration (CASTOR HSM only)
.TP
BEING_MIGR	being migrated by the automatic migrator (CASTOR HSM only)
.TP
WAITING_NS	waiting for creation in nameserver (CASTOR HSM only)
.TP
WAITING_MIGR	waiting for automatic migration request (CASTOR HSM only)
.RE
.HP
Nbaccess number of requests for this disk file.
.HP
Size	actual size / requested size (in Mbytes).
.SH OPTIONS
.TP
.BI \-A " pattern"
searches for files allocated with
.B stagealloc
and whose name matches the specified pattern.
.B stageqry
treats a pattern as a basic regular expression which should be surrounded
by single quotes to avoid unwanted shell substitutions.
.TP
.BI \-a
gives the status for all files. If this option is absent, only the files
that belong to the requestor's group are displayed.
.TP
.BI \-f
gives on an extra line the full external_filename or the full HSM filename.
.TP
.BI \-h " host"
sends the request to the stager on
.IR host .
This may also be specified thru the environment variable STAGE_HOST.
.TP
.BI \-I " external_filename"
prints the catalog entry corresponding to this external disk file.
.TP
.BI \-L
gives the link name(s) for the selected files.
.TP
.BI \-l
gives more details for each file: the login name and the group name of the
user who requested the initial stage operation, the creation date and the
date of last access.
.TP
.BI \-M " pattern"
searches in the stager catalog for HSM files whose name matches the specified
pattern.
.B stageqry
treats a pattern as a basic regular expression which should be surrounded
by single quotes to avoid unwanted shell substitutions.
.TP
.BI \-P
gives the internal path name(s) for the selected files.
.TP
.BI \-p " poolname"
restricts the query to a given pool.
.TP
.BI \-q " file_sequence_number"
restricts the search to tape files having this given sequence number.
The option value is considered as a single file sequence (case of concatenation).
It is exclusive with \-Q option.
.TP
.BI \-Q " file_sequence_range"
restricts the search to tape files having the sequence numbers within the given range.
Multiple file numbers may be separated by commas or specified as a range
using the dash as separator.
It is exclusive with \-q option.
.TP
.BI \-S
gives on stdout a list of staged files sorted in ascending order of last access
time weighted by file size.
.br
        W = max (atime, mtime) - (86400 * log (size / 1024))
.br
There are six fields per file: date of last access, time of last access, size
in Mbytes, number of accesses, internal path and user path.
.TP
.BI \-s
gives statistics on pool utilization.
.TP
.BI \-T
gives on standard output, as an option string, the main characteristics of a
tape file. The information is taken from the header labels. This includes
block size (-b), record format (-F), file identifier (-f) and record length (-L).
.TP
.BI \-u
restricts the query to files that belong to the requestor.
.TP
.BI \-V " vid(s)"
restricts the search to files corresponding to given vids.
Multiple vids will be separated by colons.
.TP
.BI \-x
adds two columns to the output: they give the request id and the internal
pathname.
.TP
.BI \--class
gives CASTOR's file classes specifications. Please note that the fileclasses specifications listed will only be those that were concerned by any file that was or is beeing migrated. In particular if a given entry is already STAGED when the stager daemon starts up and no new file, belonging to the same fileclass, appears to be or have be a candidate for migration up to when you run this stageqry command, such a fileclass will not be listed. Used only if -s option is set.
.TP
.BI \--counters
gives CASTOR's read/write counters for stageout pools. Those counters are used to select the best filesystem while doing a stageout, taking into account other running stageout's as well as running migrations.
.TP
.BI \--dump
dumps the content of the found entry(ies) in the main catalog, or in the path catalog in case of -L option.
.TP
.BI \--migrator
gives statistics on migration rules. Used only if -s option is set.
.TP
.BI \--mintime
gives current minimum time before migration. This concerns only entries that are in the CAN_BE_MIGR state.
.TP
.BI \--noregexp
prevents regular expression to be applied in case of -A of -M options. You then have to give the full (hsm) name as it was given when the entry was created inside the stager.
.TP
.BI \--queue
gives CASTOR's waiting queue content.
.TP
.BI \--reqid " reqid "
outputs only entries that have this given reqid.
.TP
.BI \--retenp
gives current retention period on disk. This period will be showed in terms of either maximum timestamp, or generic keywords that are: AS_LONG_AS_POSSIBLE and INFINITE_LIFETIME, or Exhausted. The CASTOR files lifetime can be changed using the \fBstagesetretenp\fP command. A retention period of INFINITE_LIFETIME guarantees that the corresponding entry will never be purged. A retention period of AS_LONG_AS_POSSIBLE says that the corresponding entry can be purged if there is missing space into internal disks. Another value specifies that the corresponding entry will be automatically removed if local time of the stager daemon exceeds printed value, or will be removed almost immediated if it says Exhausted.
.SH EXAMPLES
.TP
stageqry
.nf
.cs R 18
Vid    Fseq Lbl Recfm Lrecl Blksiz State      Nbaccess     Size    Pool
CZ0134    1 al  U         *  32760 STAGED            3    0.3/200  stagetest
CZ0134    2 al  U         *  32760 STAGEIN           1    0.0/200  stagetest
.cs R
.fi
.TP
stageqry -A '^MyDice' 
.nf
.cs R 18
File name                          State      Nbaccess     Size    Pool
MyDice.sav                         STAGED            2    0.2/1    stagetest
.cs R
.fi
.TP
stageqry -l
.nf
.cs R 18
Vid    Fseq Lbl Recfm Lrecl Blksiz State      Nbaccess     Size    Pool
CZ0134    1 al  U         *  32760 STAGED            3    0.3/200  stagetest
			created by  baud      c3  1994/01/12 17:54:45
			last access               1994/01/13 07:18:28
CZ0134    2 al  U         *  32760 STAGED            1    0.6/200
			created by  baud      c3  1994/01/13 07:18:28
			last access               1994/01/13 07:30:04
.cs R
.fi
.TP
stageqry -L
.nf
.cs R 18
shd02:/u4/c3/baud/SHIFT/stage/fort.41
shd02:/u4/c3/baud/SHIFT/stage/xxx
.cs R
.fi
.TP
stageqry -M run1193.raw -f
.nf
.cs R 18
File name                            State      Nbacc.     Size    Pool
run1193.raw                           STAGED         1  191.7/250  wa97_stage
 hpss1d01:/hpss/cern.ch/user/c/cdrna57/raw/1997/run1193.raw
.cs R
.fi
.TP
stageqry -P
.nf
.cs R 18
shd02:/stage/c3/stage/CZ0134.1.al
shd02:/stage/c3/stage/CZ0134.2.al
.cs R
.fi
.TP
stageqry -S
.nf
.cs R 18
1994/01/13 07:18:28    0.3    3 shd02:/stage/c3/stage/CZ0134.1.al shd02:/tmp/fort.41
.cs R
.fi
.TP
stageqry -s
.nf
.cs R 18
POOL stagetest DEFSIZE 200 MINFREE 10 GC shd02:/usr/local/bin/stage_clean
                              CAPACITY 492.00M FREE 476.39M ( 96.8%)
  shd02 /stage CAPACITY 492.00M FREE 476.39M ( 96.8%)
.cs R
.fi
.TP
stageqry -T
.nf
.cs R 18
-b 32760 -F U -f SOMEDATA -L 32760
.cs R
.fi
.SH RETURN CODES
\
.br
0	Ok.
.br
1	Bad parameter.
.br
2	System error.
.br
4	Configuration error.
.SH SEE ALSO
\fBstagechng\fP(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>



CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


castor/client/StageQry.cpp
/******************************************************************************
 *                      StageQry.cpp
 *
 * This file is part of the Castor project.
 * See http://castor.web.cern.ch/castor
 *
 * Copyright (C) 2003  CERN
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * @(#)$RCSfile: StageQry.cpp,v $ $Revision: 1.5 $ $Release$ $Date: 2004/07/29 17:34:10 $ $Author: sponcec3 $
 *
 *
 *
 * @author Sebastien Ponce
 *****************************************************************************/

// Include Files
#include <iostream>
#include <vector>
#include "castor/ObjectSet.hpp"
#include "castor/rh/File.hpp"
#include "castor/rh/StageQryRequest.hpp"
#include "castor/exception/Exception.hpp"
#include "castor/client/StageQryResponseHandler.hpp"
#include "stage_constants.h"

// Local Files
#include "StageQry.hpp"

//------------------------------------------------------------------------------
// buildRequest
//------------------------------------------------------------------------------
castor::rh::Request* castor::client::StageQry::buildRequest()
  throw (castor::exception::Exception) {
  // First reject some flags parsed by BaseCmdLineClient
  std::vector<std::string> rejected;
  rejected.push_back("K");
  rejected.push_back("A");
  rejected.push_back("silent");
  rejected.push_back("nowait");
  rejected.push_back("rdonly");
  rejectFlags(rejected, "stageqry");
  // uses some other flags
  u_signed64 flags = 0;
  setRhHost();
  setRhPort();
  std::string poolName = getPoolName();
  if (m_inputFlags.find("a") != m_inputFlags.end()) {
    flags |= STAGE_ALL;
  }
  if (m_inputFlags.find("noretry") != m_inputFlags.end()) {
    flags |= STAGE_NORETRY;
  }
  // Build request
  castor::rh::StageQryRequest* req =
    new castor::rh::StageQryRequest();
  req->setFlags(flags);
  int n = 0;
  for (std::vector<std::string>::const_iterator it = m_inputArguments.begin();
       it != m_inputArguments.end();
       it++) {
    castor::rh::File* f = new castor::rh::File();
    f->setName(*it);
    f->setPoolname(poolName);
    f->setXsize(0);
    n++;
    req->addFiles(f);
    f->setRequest(req);
  }
  return req;
}

//------------------------------------------------------------------------------
// responseHandler
//------------------------------------------------------------------------------
castor::client::IResponseHandler*
castor::client::StageQry::responseHandler() throw() {
  return new StageQryResponseHandler();
}

//------------------------------------------------------------------------------
// usage
//------------------------------------------------------------------------------
void castor::client::StageQry::usage(std::string error) throw() {
  std::cout << error << std::endl;
  std::cout << "usage : stageqry [-a] [-h rh_host] [-p poolname] [--noretry] hsmfile..."
            << std::endl;
}

//------------------------------------------------------------------------------
// main
//------------------------------------------------------------------------------
int main(int argc, char** argv) {
  castor::client::StageQry req;
  req.run(argc, argv);
  return 0;
}
